### PR TITLE
Buttons for forkphorus and turbowarp

### LINKED ISSUE #79: [New addon] Quick launch to various Scratch compilers/players (Phos., Fork., Leopard, etc)
**Describe the new addon**

A way to quickly launch projects to Phosphorus, Sulfurous, Forkphorus, Leopard, and other various Scratch compilers/players (if that's the right word.)

### PR DESCRIPTION
**Pull request by NoobTracker**
Resolves #79
This PR adds a new extension - one that adds little buttons for forkphorus and other third-party websites. viewon
It just takes the URL and searches for the last truthy part. If there's a better way, please tell me!

I set it to be on by default since imo that makes sense - if you don't think so, let me know and I set it to disabled by default.
Currently there's only the ability to use forkphorus and turbowarp, I'll add support for anything as soon as lists are possible how I described them here (https://github.com/ScratchAddons/ScratchAddons/issues/142#issuecomment-684932022), until then I see no reason for adding an input for custom websites. When they're good enough then I can just add them for everyone I guess.
turbowarp is disabled by default since it's imo the worse platform. forkphorus has a better fullscreen, a dark mode, less loading time and about the same speed.

### CODE REVIEW

**Comment by NoobTracker** [comment]
Ouch. I used prettier (just by changing the lenght per line setting to 120) but I didn't with the JSON.

**Comment by NoobTracker** [comment]
Uh for some reason prettier didn't change anything at all ... can someone help me?

**Comment by NoobTracker** [comment]
Arrrrrgh, it would be nice if GitHub could give me a more precise error .... I used prettier, of course, I wrote the script in the simple text editor so I kinda didn't care about propper formatting. *sigh*

**Comment by WorldLanguages** [comment]
@NoobTracker Enable GitHub Actions in your fork, then make a dumb commit "add a space somewhere". It will lint automatically.

File: addons/third-party-button/addon.json
Comment (WorldLanguages) [inline addons/third-party-button/addon.json:10]: It's cool that you used options, but are they really necessary?

File: addons/third-party-button/userscript.js
Comment (WorldLanguages) [inline addons/third-party-button/userscript.js:11]: Interesting way to go through an URL lol
I would have just used `new URL(location.href).pathname.split("/")[2]`

**Comment by YOYITsM3M8** [comment]
I have a userscript that does exactly this but only for TurboWarp @yoyitsm3m8/userscripts

**Comment by JeffaloBob** [comment]
i dont know if this is possible but could the actual player on the page be swapped out? or have the new player embedded over top of it?

**Comment by YOYITsM3M8** [comment]
> i dont know if this is possible but could the actual player on the page be swapped out? or have the new player embedded over top of it?

It should be possible to embed the TurboWarp full screen over the actual player

**Comment by JeffaloBob** [comment]
also that misalignment is driving me mad lol

**Comment by YOYITsM3M8** [comment]
> also that misalignment is driving me mad lol

xD I just added a button with the same code and styles except obviously red and linked to TurboWarp

**Comment by WorldLanguages** [comment]
> i dont know if this is possible but could the actual player on the page be swapped out? or have the new player embedded over top of it?

Yeah, that would be very cool. I think it would work, as long as these players allow embeds. Also, cloud data probably wouldn't work?
It would also be nice to stop the real Scratch from loading, if it's not the default player choice. If the user wants to use Scratch, they must refresh the page (unless there's some magic way to revive the editor after we stopped it from loading)

File: addons/third-party-button/userscript.js
Comment (NoobTracker) [reply addons/third-party-button/userscript.js:11]: well you can use /projects/id instead of /projects/id/ and that would screw everything up so I just look for the last truthy part.

File: addons/third-party-button/addon.json
Comment (NoobTracker) [reply addons/third-party-button/addon.json:10]: I think that having them is neat so I really dunno why not. The "line sprite effect" as I call it: A feature that isn't really useful, but it would just be efford to remove it so why bother.

**Comment by NoobTracker** [comment]
That is a really neat idea, I gotta say! Wow ... you could even have the forkphorus player _in the editor_, like something I suggested a while ago! I'm gonna look into that forkphorus embedding stuff, but for now I think that the links are already pretty useful.

**Comment by NoobTracker** [comment]
Yay, now it works.

**Comment by NoobTracker** [comment]
Just a quick note: It seems like the script fails in one out of 50 times - I really dunno why.

**Comment by NoobTracker** [comment]
Oh, and for some reason the button is gone when you go into the editor and back ... strange .... has someone an idea why?

**Comment by YOYITsM3M8** [comment]
image
lol i replace the scratch stage with a iframe to turbowarp fullscreen

**Comment by NoobTracker** [comment]
Good idea! How do you get the project ID? (#158)

**Comment by YOYITsM3M8** [comment]
var current_url = document.location; // get url
var id = current_url.href.split('/')[4]; // get project id from url

**Comment by NoobTracker** [comment]
Ah so more or less the way I do it too.

**Comment by YOYITsM3M8** [comment]
Yes this way it grabs project id no mater what

**Comment by NoobTracker** [comment]
kinda yeah

**Comment by NoobTracker** [comment]
take a look at this: https://github.com/ScratchAddons/ScratchAddons/issues/158#issuecomment-685691543

**Comment by YOYITsM3M8** [comment]
Or one line 

id = document.location.href.split(“/“)[4]

**Comment by NoobTracker** [comment]
Well I guess you can do that too but there in the comment is a better way of doing it.

**Comment by YOYITsM3M8** [comment]
Yes but I use my one liner for Userscripts

**Comment by NoobTracker** [comment]
Maybe you can stop using userscripts with this extension.

**Comment by YOYITsM3M8** [comment]
Lol, my Userscript are custom coded like “add turbo warp button to project page and editor”
“Replace player with turbo warp”

**Comment by YOYITsM3M8** [comment]
Etc.

**Comment by NoobTracker** [comment]
> 
> 
> Lol, my Userscript are custom coded like “add turbo warp button to project page and editor”
> “Replace player with turbo warp”

I think that's exactly the kind of userscript we want in this extension.

**Comment by NoobTracker** [comment]
Anyway, 


> 
> 
> Just a quick note: It seems like the script fails in one out of 50 times - I really dunno why.



> 
> 
> Oh, and for some reason the button is gone when you go into the editor and back ... strange .... has someone an idea why?

Those are currently problems, has anyone a clue how to solve them?

**Comment by YOYITsM3M8** [comment]
> > Lol, my Userscript are custom coded like “add turbo warp button to project page and editor”
> > “Replace player with turbo warp”
> 
> I think that's exactly the kind of userscript we want in this extension.

https://github.com/YOYITsM3M8/userscripts/blob/master/TurbowarpScratchMod.user.js

**Comment by NoobTracker** [comment]
Yeah you can port it pretty easily I think. There's somewhere a guide, linked in the readme.

**Comment by NoobTracker** [comment]
For some reason it fails quite often what results in a `Uncaught (in promise) TypeError: document.getElementsByClassName(...)[0] is undefined` error. It seems like it can't find an object with the class name "project-title". Weird ....

**Comment by NoobTracker** [comment]
> 
> 
> also that misalignment is driving me mad lol

wdym? If you think that the buttons aren't placed properly, feel free to make a quick PR adding some CSS.

File: addons/third-party-button/userscript.js
Comment (apple502j) [inline addons/third-party-button/userscript.js:13]: Please do not use innerHTML, as it may create security risks.

Comment (NoobTracker) [reply addons/third-party-button/userscript.js:13]: Huh?! Okay, what else should I do?

**Review by Hans5958 — APPROVED** [review]
40th comment

File: addons/third-party-button/userscript.js
Comment (Hans5958) [reply addons/third-party-button/userscript.js:13]: `textContent`, but I'm not sure if it's possible to insert elements with `textContent`. If not, we can take the scenic route.

**Comment by NoobTracker** [comment]
Congratulations, @Hans5958

**Comment by YOYITsM3M8** [comment]
Can someone tell me what’s wrong with my PR? #164

File: addons/third-party-button/userscript.js
Comment (NoobTracker) [reply addons/third-party-button/userscript.js:13]: Yeah I already thought about the fact that you *could* have a URL containing `"><script>/*evil*/</script><"` or something but since you can't enter one and since it would also be kinda stupid, it's not really a risk.

**Comment by Hans5958** [comment]
> 
> 
> Can someone tell me what’s wrong with my PR? #164

I actually misunderstand. Left a comment on #164.

**Comment by NoobTracker** [comment]
@YOYITsM3M8 It's that @Hans5958 thinks that these two things should merge since they're both essentially just an easier way of using fork and stuff.

File: addons/third-party-button/userscript.js
Comment (Hans5958) [reply addons/third-party-button/userscript.js:13]: Yeah, it is valid if it's done in a XSS, but, man, this is far from one.

**Comment by YOYITsM3M8** [comment]
Ah, k

**Comment by YOYITsM3M8** [comment]
@NoobTracker we can merge these to one and make it so you can toggle

**Comment by NoobTracker** [comment]
@YOYITsM3M8 I think that it would be better to not just replace the stage and not just in the editor. I think I'm gonna make a PR based on yours.

**Comment by YOYITsM3M8** [comment]
Kkk

**Comment by NoobTracker** [comment]
Has anyone an idea why this happens and how to prevent it?


> 
> 
> For some reason it fails quite often what results in a `Uncaught (in promise) TypeError: document.getElementsByClassName(...)[0] is undefined` error. It seems like it can't find an object with the class name "project-title". Weird ....

Can anyone test that?

**Comment by YOYITsM3M8** [comment]
I know how, in my code it has a script that waits until that element is created bc it is not in the html other js codes add it

**Comment by NoobTracker** [comment]
Okay that does that. But is there a better way then checking every 300ms if it exists?

**Comment by YOYITsM3M8** [comment]
Using jquery I think

**Comment by NoobTracker** [comment]
Ummmmm, can anyone tell me how exactly that works? I really have _no_ clue how jquery works.

**Comment by YOYITsM3M8** [comment]
Just google "wait until element exists jquery" or something like that

**Comment by NoobTracker** [comment]
I can't find a command for that ...

**Comment by NoobTracker** [comment]
They all just rely on checking something every x ms.

**Comment by NoobTracker** [comment]
@WorldLanguages @JeffaloBob Do you know a better way to wait for an element then just checking it every x ms?

**Comment by YOYITsM3M8** [comment]
nope

**Comment by NoobTracker** [comment]
@YOYITsM3M8 wdym?

**Comment by YOYITsM3M8** [comment]
I don’t know any other ways

**Comment by NoobTracker** [comment]
... and the other people I mentioned?

Also, why is the project title created by a script?

**Comment by YOYITsM3M8** [comment]
Idk

**Comment by JeffaloBob** [comment]
the placement of the links is pretty awkward...

**Comment by YOYITsM3M8** [comment]
Yeah I agree

**Comment by NoobTracker** [comment]
@JeffaloBob @YOYITsM3M8 where should I place it? I can move it I guess ...

**Comment by YOYITsM3M8** [comment]
Just align on center

**Comment by YOYITsM3M8** [comment]
Not top

**Comment by JeffaloBob** [comment]
how about 
image (terrible mockup but you get the idea)


then the players thing could dropdown links to different players.

i think having text next to the see inside button looks kinda wack in my opinion and this button might make it a bit cleamer

**Comment by NoobTracker** [comment]
Well there it could actually be, maybe just two buttons with simply "forkphorus" and "turbowarp" on them?

**Comment by YOYITsM3M8** [comment]
Yes but with colors

**Comment by YOYITsM3M8** [comment]
Red for turbo warp and black for fork

**Comment by JeffaloBob** [comment]
> Well there it could actually be, maybe just two buttons with simply "forkphorus" and "turbowarp" on them?

i dont think that would look good because then you just end up bloating the interface with more buttons(?)



> Red for turbo warp and black for fork

scratch has such a nice matching colour scheme, adding more colours would just make it a bit busy though...

**Comment by NoobTracker** [comment]
We'll see. We could make many stuff optional.
But gtg c u in 10 minutes or something

**Comment by YOYITsM3M8** [comment]
But then they pop out so u could actually see them (it took me a bit to find the difference on ur mockup)

**Comment by NoobTracker** [comment]
Just one or two buttons would be easier and more useful i guess

**Comment by NoobTracker** [comment]
since the main goal of this extension is making using fork and stuff easier, it isn't very helpful to hide those buttons.

**Comment by WorldLanguages** [comment]
> @WorldLanguages @JeffaloBob Do you know a better way to wait for an element then just checking it every x ms?

`addon.tab.waitForElement`: 
https://github.com/ScratchAddons/ScratchAddons/wiki/addon.*-APIs-reference#addontabwaitforelementselector

**Comment by NoobTracker** [comment]
@WorldLanguages Thanks! What exactly does that function?

**Comment by WorldLanguages** [comment]
Returns a promise that resolves when the specified element is found. You can use it together with `await` to have some nice code that depends on a DOM element being present.

**Comment by NoobTracker** [comment]
Thanks! I'm also gonna make the links nicer ...

**Comment by NoobTracker** [comment]
Um @WorldLanguages can you give me an example how to use it?

**Comment by WorldLanguages** [comment]
```js
await addon.tab.waitForElement(".elementSelector");
console.log("Element exists!: ", document.querySelector(".elementSelector");
```

**Comment by NoobTracker** [comment]
... why doesn't that run prettier _before_ it checks it the first time?

**Comment by NoobTracker** [comment]
Ah okay so it basically stops everything until that is loaded.

**Comment by WorldLanguages** [comment]
Yep, what element are you waiting for? You need to get a selector for it (I think it supports any CSS selector)

**Comment by NoobTracker** [comment]
I already added that line :)
Yeah, it's like document.querySelector (which I never used before btw) so it takes CSS selectors

**Comment by NoobTracker** [comment]
So now we can _merge_ it ...

... besides a small problem: Those buttons disappear when you go into the editor and back, don't ask me why! But I think that it can already be merged as Beta, I'm gonna continue work (whoops, almost wrote fork) tomorrow. Good night or something because it's probably afternoon where you live but anyway bye.

**Comment by apple502j** [comment]
Please change the code to remove any use of innerHTML.

**Comment by Explosion-Scratch** [comment]
:O 91 comments =P

**Comment by YOYITsM3M8** [comment]
Can you send a pic of what it looks like now

File: addons/third-party-button/userscript.js
Comment (GarboMuffin) [inline addons/third-party-button/userscript.js:24]: use #ff4c4c

File: addons/third-party-button/userscript.js
Comment (GarboMuffin) [inline addons/third-party-button/userscript.js:23]: use #1f313e

**Comment by WorldLanguages** [comment]
> Those buttons disappear when you go into the editor and back
Yeah, Griffpatch's developer tools also do something weird because apparently, all the GUI is reconstructed each time you go inside and outside the editor.

File: addons/third-party-button/userscript.js
Comment (NoobTracker) [reply addons/third-party-button/userscript.js:24]: Ah thank you I'll change that.

File: addons/third-party-button/userscript.js
Comment (NoobTracker) [reply addons/third-party-button/userscript.js:23]: Okay ...

**Comment by NoobTracker** [comment]
> 
> 
> > Those buttons disappear when you go into the editor and back
> 
> Yeah, Griffpatch's developer tools also do something weird because apparently, all the GUI is reconstructed each time you go inside and outside the editor.

Yeah it seems like that's the case.

**Comment by NoobTracker** [comment]
> 
> 
> Please change the code to remove any use of innerHTML.

How exactly?

**Comment by NoobTracker** [comment]
fork
Those are the old colors (@JeffaloBob you can turn the colors off if you want) since I'm too lazy to update them on my version but that doesn't matter since it should just show how it looks.

**Comment by Hans5958** [comment]
We surpassed the 100 comment mark. It's pretty rare to have a issue/PR with more than 100 comments in it.

**Comment by NoobTracker** [comment]
Well that just means that we _really_ want to have a good extension :P

**Comment by Hans5958** [comment]
> 
> 
> Well that just means that we _really_ want to have a good extension :P

It's more like you guys seperate your comments that can be sent with only one, with edits.

**Comment by NoobTracker** [comment]
Well I guess that's also true ....
... but we also want a _really_ good extension :P

**Comment by NoobTracker** [comment]
What exactly is the problem with innerHTML?

**Comment by Hans5958** [comment]
Yeah, something something #173.

**Comment by Hans5958** [comment]
> 
> 
> Okay ... I guess I can do it the other more complicated way that makes no sense imo but who cares after I made a quick change ...

`textContent` exists, though.

**Comment by NoobTracker** [comment]
But there's a div inside the link.

**Comment by NoobTracker** [comment]
So that's annoying.

**Comment by Hans5958** [comment]
1 message pls
Nvm, I thought you wanted to do text only.

**Comment by NoobTracker** [comment]
Anyway ...
I found out that `onClick` doesn't result in an event when you use the mouse wheel. Sooooo .... that means that I can add some stuff so you click on the button and the stage gets replaced but you use your mouse wheel and you get a new tab! Of course, that would be optional. Do you think that it should be enabled by default?

**Comment by NoobTracker** [comment]
Okay! That kinda seems to work now! I can click on the button and the stage is replaced by fork.

**Comment by NoobTracker** [comment]
So now I have a problem: I want to make a function for replacing the stage and stuff but I just get errors.
Can you tell me how I can use `e` as a parameter? `fork.addEventListener("click", function(e){e.preventDefault();});` works but not
```
function doStuff(e){e.preventDefault();}
fork.addEventListener("click", doStuff(e));
```
and I don't understand why and what I do wrong.

**Comment by NoobTracker** [comment]
Ah okay that does that! Thank you very much!

**Comment by NoobTracker** [comment]
@Hans5958 Uhhhh ... do you have an idea how to make fullscreen possible?
Hold on, I quickly update the stuff that works.

**Comment by NoobTracker** [comment]
> 
> 
> Also, cloud data probably wouldn't work?

Forkphorus is currently working on cloud variables - they have their own server and it's really fast - 15 refreshes per second and not such a small size limit!

**Comment by NoobTracker** [comment]
Also, does anyone know the exact width and height values for TurboWarp? @YOYITsM3M8 ?

**Comment by NoobTracker** [comment]
Ummm I currently redo pretty much everything is it okay when I just reupload the file and don't try to recreate exactly what I did? It would be easier but less organized so you decide.

**Comment by YOYITsM3M8** [comment]
Yes I do

**Comment by NoobTracker** [comment]
... okay? So what are the exact vaules?

**Comment by YOYITsM3M8** [comment]
width="482.22"height="406.22"; height="406.22"; height="406.22"; height="406.22";

**Comment by YOYITsM3M8** [comment]
Ahhh to much paste

**Comment by NoobTracker** [comment]
ah thx

**Comment by NoobTracker** [comment]
> 
> 
> Ummm I currently redo pretty much everything is it okay when I just reupload the file and don't try to recreate exactly what I did? It would be easier but less organized so you decide.

**Comment by YOYITsM3M8** [comment]
..?

**Comment by NoobTracker** [comment]
@WorldLanguages Ooops I rerequested a review accidentaly sorry ...

**Comment by NoobTracker** [comment]
Ah btw there is another option that's neat I think: When you click the fork button the first time, it just replaces the stage, but when you click it when the stage is already replaced, then it acts like a normal link and you get redirected to the fork website. You can simply double-click. 
But that's just an option and it's turned off by default.

**Comment by YOYITsM3M8** [comment]
add that for turbo

**Comment by Hans5958** [comment]
Just an idea, I really wanted to implement a dropdown menu instead of buttons for other players like Phosphorus, Leopard, and some players that might be released in the future.

Also, I noticed the name "Third party button." Like, third party, what? How about something like "Third party player button" or "External player button?"

**Comment by NoobTracker** [comment]
External player button is a good idea. Let me rename that.

Also, do you really want to have 5 different players? I don't. I would just use forkphorus. Some people might want to have also turbowarp or something else, but you don't have to enable them all, right?

**Comment by NoobTracker** [comment]
> 
> 
> Add a option for which player

wdym?

**Comment by NoobTracker** [comment]
Uhhhhhh how can I rename a folder on GitHub?

**Comment by NoobTracker** [comment]
Those are currently the options:
options

**Comment by Hans5958** [comment]
> Also, do you really want to have 5 different players?

Yes, beacause why not?

> Some people might want to have also turbowarp or something else, but you don't have to enable them all, right?

You underestimated the users. I knew someone that adds around 600 "things" in a single application, so I expect there are some people who will turn on all addons and options possible.

**Comment by NoobTracker** [comment]
Well ... if there were things like lists and scripts that run when you want to change options then I could probably make something cool possible ...
But for now, I might just add the option to add an option to a dropdown instead of just creating a button for it.


> 
> 
> You underestimated the users. I knew someone that adds around 600 "things" in a single application, so I expect there are some people who will turn on all addons and options possible.

Well, currently it's only for Warpand fork and I might add an "Other" text input, that's only three buttons. And really, I think that it just doesn't make that sense to enable everything.

**Comment by NoobTracker** [comment]
As I said: The main point of this extension is making using forkphorus and other stuff easier. When you first have to click on a dropdown and then find your desired player in a list of 20, then the main point is gone.

**Comment by NoobTracker** [comment]
Wow am I completely blind or is there just not a button for renaming a folder?

**Comment by YOYITsM3M8** [comment]
There isn’t on the website

**Comment by NoobTracker** [comment]
On the website you have the fork and the turbo button so you can select which player you want. By default it's just Scratch.

**Comment by YOYITsM3M8** [comment]
i am talking about renaming a folder

**Comment by NoobTracker** [comment]
Ah how awesome. So how do I do that?

**Comment by YOYITsM3M8** [comment]
I mean u cant do it on the website I don’t think

**Comment by YOYITsM3M8** [comment]
So oof

**Comment by NoobTracker** [comment]
... if not on the website, where do I do it?

**Comment by YOYITsM3M8** [comment]
Idk git cli or github desktop

**Comment by NoobTracker** [comment]
Okay now that should work.

**Comment by Hans5958** [comment]
> Well ... if there were things like lists and scripts that run when you want to change options then I could probably make something cool possible

https://www.w3schools.com/jsref/event_onchange.asp, probably?

> As I said: The main point of this extension is making using forkphorus and other stuff easier. When you first have to click on a dropdown and then find your desired player in a list of 20, then the main point is gone.

Sure, depending on how you look at it. I look it in a different approach

**Comment by NoobTracker** [comment]
Okay now I'm finally kinda happy with the current state. It can be merged imo.

**Comment by NoobTracker** [comment]
@Hans5958 I mean the possibility to mess with the extension's setting page so I can say that one option is only there when a boolean is activated and such stuff.

**Comment by Hans5958** [comment]
Ah.

File: addons/external-player-button/userscript.js
Comment (apple502j) [inline addons/external-player-button/userscript.js:2]: Is this necessary?

Comment (NoobTracker) [reply addons/external-player-button/userscript.js:2]: I dunno, I just copied another extension's stuff and kept it for ... um ... reasons.

Comment (YOYITsM3M8) [reply addons/external-player-button/userscript.js:2]: XD

Comment (NoobTracker) [reply addons/external-player-button/userscript.js:2]: For quite obvious reasons which are ... umm ... 
Next question!

**Comment by NoobTracker** [comment]
Just a little question: Does it make sense to add the instantly running thing in the JSON?

**Comment by apple502j** [comment]
Try it yourself - if it works, maybe?

**Comment by NoobTracker** [comment]
Well currenly I don't have that update in my fork so I'll do that after the merge.

**Comment by NoobTracker** [comment]
So, can it be merged now?

**Comment by apple502j** [comment]
Waiting for final review from W_L

**Comment by WorldLanguages** [comment]
Looks good

**Review by WorldLanguages — APPROVED** [review]

**Comment by NoobTracker** [comment]
Yay! So now it can be merged?

File: addons/external-player-button/addon.json
Comment (WorldLanguages) [inline addons/external-player-button/addon.json:48]: Just set this to false. And tags to "community"

Comment (NoobTracker) [reply addons/external-player-button/addon.json:48]: Ah okay

Comment (WorldLanguages) [reply addons/external-player-button/addon.json]: Remove "projects".

Comment (NoobTracker) [reply addons/external-player-button/addon.json]: Um doesn't it mess with the project pages?

Comment (NoobTracker) [reply addons/external-player-button/addon.json]: Whatever

**Comment by NoobTracker** [comment]
If there's another thing that isn't right so far it would be nice if you could tell me in the next 3 minutes or so because it's late here and my parents want that I shut my PC down.

**Comment by NoobTracker** [comment]
Phewwwwww [...] wwww

**Comment by Hans5958** [comment]
PR closed with 171 comments. Wow.

**Comment by NoobTracker** [comment]
Well we also kinda used this PR as a chat and I really commented everything I just thought about but well .... I don't care. It's now done.

**Pull request merged**
